### ISSUE TITLE
[RFC] Building plugins as system packages instead of shipping shell scripts

### ISSUE THREAD

**Issue by sudoforge** [fedy/fedy#448]
# Overview

Currently, the majority of plugins supported by Fedy and the Folks With Hats team are shipped as shell scripts, utilizing the `install.sh`, `uninstall.sh`, and `metadata.json` scripts. As a result of this decision, there are several points of pain with the current application workflow.

Note that in regards to #422, #447, this proposal should be considered made for `officially supported plugins` only. Community-built plugins are still a WIP, and we do not have a specific plan for implementation.

## Definitions

To start, let's define a few terms so that everyone is on the same page when discussing this topic.

- `target` is the resulting application, tool, or other item(s) that ends up on the user's system, after installing dependencies and configuration is completed.
- `plugin` means the bundle of commands/scripts used by Fedy to install a `target`.
- `package` is an RPM package, built from a `spec` file.

## Problems with scripts

There are several problems with the current way we ship plugins.

### Confusing upgrade steps

Once a user installs a target through Fedy's plugin workflow, there is no way to update the steps defined within the plugin. If we need to ship changes to a user, like say, a refactored installation step or new version of the target, they need to:

- uninstall the target by clicking the `Remove` button
- update the `fedy` application via `dnf up{date,grade}` to get the latest bundled versions of the plugin's scripts
- reinstall the target by clicking the `Install` button.

Additionally, these are the same steps a user would have to take in order to update a target that Fedy installed, but which doesn't have a built-in method for updating.

### Unmanaged versions

Many plugins install the latest version of the target from the target's origin source. This can be convenient, but allows for inconsistencies between users of Fedy, which can sometimes be the source of a bug, and can certainly cause issues when trying to replicate an issue. If a user installed a target from a plugin in January, that target could receive weekly updates, and the user would be unaware. They might report a bug six months down the road, and due to unmanaged versions of targets, it may be more difficult to track down that bug.

## Pros of using system packages 

- All plugins are easily manageable, and updates can be shipped to users via their next system update
- Targets can be locked into specific versions, which will ensure that all plugins of a specific release will always work as expected
- Simpler file tree: a single `spec` file can be used to define all attributes and (un/in)stallation steps currently split across `install.sh`, `uninstall.sh`, and `metadata.json`.
- Easy installation via command line. Users will be able to install any Fedy plugin through the command line, for example - `dnf install fedy-plugin-chrome`. This makes workstation replication easier to automate, with much less boilerplate than copying the existing plugin schemas. Abstracting plugin installation from the GUI also makes our ecosystem portable to other systems.
- Easy, consistent replication. Within each release, every user will always have the same plugins and versioned targets, so bugs are easy to replicate.
- Dependency checks and caching is handled through the system package manager (in Fedora's case, `dnf`).

## Cons of using system packages

- Contributing becomes slightly more complex, as the `spec` file is a structured document, and while relatively easy to reason about, is not as straightforward as a shell script written in `bash`. This might end up raising the barrier to contribution to the official library, although this proposal argues that proper contributing documentation will actually _lower_ the barrier to entry from what it is today.

- Slightly more overhead when creating a new release, due to the requirement of building packages for all plugins. This will be abstracted away to an automated, simple-to-execute workflow (one command, no need to install build tools), so this proposal does not see this as an issue.

## Additional notes

- The process of handling release versions and building the packages will be entirely automated. This means that the only thing required to add a new plugin to Fedy, in this proposed workflow, will be to create the `spec` file.
- When a `target` provides its own method of updating, or provides its own repository files, the package provided by the Fedy plugin will simply add or enable the required repository and install using the system package manager -- like Visual Studio Code, or Google Chrome, for example. This might seem like too little work to validate building a system package, but this proposal argues that even simple plugins should be packaged so that they can gain the benefits described above.

**Comment by satya164** [comment]
Are you planning to package the plugins separately or the software installed by the plugins? For example is it `fedy-plugin-chrome.rpm` or `chrome.rpm`?

**Comment by bddenhartog** [comment]
@satya164 
> Are you planning to package the plugins separately or the software installed by the plugins? For example is it `fedy-plugin-chrome.rpm` or `chrome.rpm`?

The final note covers this. For software that has official repositories, the Fedy plugin would set up the official repository if it doesn't exist, in addition to performing any other configuring that is needed. Post installation, the Fedy plugin would install the software from the official repository.

The proposed name for plugin system packages is `fedy-plugin-*`, where the `*` is the plugin's name.

**Comment by satya164** [comment]
The name `fedy-plugin-*` sounds misleading since it's not a plugin for fedy.

**Comment by bddenhartog** [comment]
@satya164 
> The name fedy-plugin-* sounds misleading since it's not a plugin for Fedy.

It's a Fedy plugin; just like the objects in the `plugins/` directory aren't plugins for Fedy, but rather what Fedy calls a "plugin": a bundle of things that install another thing. 

We currently ship `fedy-font-config` and `fedy-multimedia-codecs`, neither of which are named appropriately, as they aren't installing _our_ font configuration or _our_ multimedia codecs.

That being said, however, I'm totally open to another naming scheme.

**Comment by bddenhartog** [comment]
@satya164 
> The name `fedy-plugin-*` sounds misleading since it's not a plugin for Fedy.

It's a Fedy plugin; just like the objects in the `plugins/` directory aren't plugins for Fedy, but rather what Fedy calls a "plugin": a bundle of things that install another thing. 

We currently ship `fedy-font-config` and `fedy-multimedia-codecs`, neither of which are named appropriately, as they aren't installing _our_ font configuration or _our_ multimedia codecs.

That being said, however, I'm totally open to another naming scheme.

**Comment by satya164** [comment]
> objects in the plugins/ directory aren't plugins for Fedy, but rather what Fedy calls a "plugin": a bundle of things that install another thing.

I don't see the difference, they **are** plugins for fedy

> We currently ship fedy-font-config and fedy-multimedia-codecs, neither of which are named appropriately, as they aren't installing our fonts or our multimedia codecs.

They are font config and multimedia codec packages **for** fedy

**Comment by bddenhartog** [comment]
Sure, but doesn't `fedy-chrome` infer that it's Fedy's "Chrome"? I'm not sure that naming scheme fits for plugins.

**Comment by bddenhartog** [comment]
Also, your two posts are contradictory. Can you please clarify what you see wrong with `fedy-plugin-*`?

> The name `fedy-plugin-*` sounds misleading since it's not a plugin for fedy.

and then....

> 
> > objects in the plugins/ directory aren't plugins for Fedy, but rather what Fedy calls a "plugin": a bundle of things that install another thing.
> 
> I don't see the difference, they **are** plugins for fedy
> 
> > We currently ship fedy-font-config and fedy-multimedia-codecs, neither of which are named appropriately, as they aren't installing our fonts or our multimedia codecs.
> 
> They are font config and multimedia codec packages **for** fedy

So something named `fedy-plugin-*` would be a plugin **for** Fedy, by following the logic in your second post.

**Comment by satya164** [comment]
Again, it depends on what you package. `fedy-plugin-chrome` will be the plugin which contains the instructions for fedy regarding how to install Google Chrome. Google Chrome, on the other hand, is not a Fedy plugin, it's a third-party app.

The font config and multimedia packages are just packages. They don't have instructions for Fedy. They are not plugins for Fedy, rather some packages that were made for Fedy.

**Comment by Abhinav1217** [comment]
I prefer current setup. Shell scripts in plugin works. There is no performance issue. Changing core plugin won't be a problem. It gives more control over commands and there is no reinventing the wheel. Fedy's main goal is to make dealing with fedora easier for users. And it's simplicity is what it is top recommended for fedora.

**Comment by Abhinav1217** [comment]
Working on community plugin will make this easier. For example, instead of bundling the core plugins with the app, We can have them downloaded via same api we would use to download community plugins.

**Comment by Abhinav1217** [comment]
Here is my opinion breakdown 

**Problems with scripts**

- **Confusing upgrade steps**

> there is no way to update the steps defined within the plugin. If we need to ship changes to a user, like say, a refactored installation step or new version of the target, they need to:
> -   uninstall the target by clicking the Remove button
> -   update the fedy application via dnf up{date,grade}
>  to get the latest bundled versions of the plugin's scripts
> -   reinstall the target by clicking the Install button.

I don't thing this is the case.
We have 2 type of plugins, 1 which uses DNF  to install softwares, DNF will take care of its updates once copr or repo url is installs. Your steps does not apply here.

2 Plugin which installs softwares by manual downloading and setting up here. In that case your steps are the only way. But you got the steps wrong. 
You don't really need to update fedy. Because we prefer not to have hard coded versioned downloaded links but regex generated url which fetches the latest download version. So it is just clicking remove button on fedy and then clicking install button on fedy. As long as the download links are valid it will install latest version of software irrespective of whether fedy is updated or not.

There are some ways we can further improve this workflow.
- Always prefer using copr or repos if they are available. Reject plugin which try to install some software directly even when they have dnf streams. Dnf will take care of updates of that software.

We can't do anything when there is no dnf means legally like in case of sublime text which does not provide repo stream nor allows you to create copr yourself. For such software, only option is if that software have self update mechanism or uninstall and reinstall newer version.

- We should enable fedy drpm. So that update size would be small and we can push frequent updates. Like if only one plugin is changed a tiny few bytes of update can be pushed.

- Have some inbuilt mechanism which would update plugins via API.

**Unmanaged versions**
Your point is covered above. We do not accept hard coded urls. Regex pattern have suggested by satya whenever someone sends PR with this issue. And I have checked every updated URL pattern in plugins. 
If the website has changed the download URL, then surely we do need to update the plugin but that is a small update. If they do not provide dnf streams or internal software updates then there is very little we can do.

**Comment by Abhinav1217** [comment]
In reading pros and cons section, Well I need to add that what you are suggesting is already done within gnome-softwares.  Writing bash script with commands will be simpler for contributers than writing spec file. For example I have never work with makefile and I ask satya to update the makefile when needed. All 3 files `install.sh, uninstall.sh, metadata.json` can be merged in one metadata.json file but It makes reviewing it more complicates. Unix philosophy, One file should do one thing well. **there is nothing wrong with divide and conquer.**   
Look at this way, If problem/bug arises in install script, we can focus on it. If it arises during uninstallation ( like we are working on managing coprs ) then just focus on uninstall script.

**Comment by bddenhartog** [comment]
@satya164 
> Again, it depends on what you package. fedy-plugin-chrome will be the plugin which contains the instructions for Fedy regarding how to install Google Chrome. A Fedy plugin is any folder which has a metadata.json with a format Fedy understands, regardless of what else it contains. Google Chrome, on the other hand, is not a Fedy plugin, it's just a third-party app.
>
>The font config and multimedia packages are just packages. They don't have instructions for Fedy. They are not plugins for Fedy, rather some packages that were made for Fedy. The fedy-* in their name indicates their association with Fedy, because they were made specifically for using with fedy/or you can say made by the fedy team. The name doesn't indicate what the association is. fedy-plugin-* makes it clear that what the type of association is.

Using the definitions from the original issue post, the `plugin` is the set of scripts Fedy uses to install (or otherwise set up/configure) the `target`, which may be a software application, a set of scripts, or in the context of [codecs.plugin](https://github.com/folkswithhats/fedy/tree/master/plugins/codecs.plugin) and [fontconfig.plugin](https://github.com/folkswithhats/fedy/tree/master/plugins/fontconfig.plugin), the list of packages to install that comprise the "thing(s) fedy set up for you as part of this fedy plugin". This seems to coincide with your idea of a "Fedy plugin".

Using Chrome as an example, `fedy-plugin-chrome` would do what I described in the original issue post:

> When a `target` provides its own method of updating, or provides its own repository files, the package provided by the Fedy plugin will simply add or enable the required repository and install using the system package manager -- take Visual Studio Code or Google Chrome, for example. This might seem like too little work to validate building a system package, but this proposal argues that even simple plugins should be packaged so that they can gain the benefits described above.

And `fedy-plugin-codecs` would simply install the necessary dependencies that comprise the "fedy plugin" for codecs.

**Comment by bddenhartog** [comment]
@Abhinav1217 _(emphasis mine)_

> I prefer current setup. Shell scripts in plugin works. There is no performance issue. Changing core plugin won't be a problem. _**It gives more control over commands and there is no reinventing the wheel**_. Fedy's main goal is to make dealing with fedora _**easier for users**_. And it's simplicity is why it is top recommended for fedora.
>
> Working on community plugin will make this easier. For example, instead of bundling the core plugins with the app, We can have them downloaded via same api we would use to download community plugins.

Fedy actually reinvents the wheel quite a bit, and has several problems:

- We had to implement our own caching system
- We are currently discussing methods of dependency management
- We have several users who run into issues using workarounds Fedy has implemented, like `rm` inconsistencies, additional prompts for installation
- We have our own, custom solution for plugin structure, with no formal documentation or guidelines

As far as ease of use goes, it is much, _much_

These issues are all resolved by using system packages for all plugins.

---

> 2) Plugin which installs softwares by manual downloading and setting up here. In that case your steps are the only way. But you got the steps wrong.
>
> You don't really need to update fedy. Because we prefer not to have hard coded versioned downloaded links but regex generated url which fetches the latest download version. So it is just clicking remove button on fedy and then clicking install button on fedy. As long as the download links are valid it will install latest version of software irrespective of whether fedy is updated or not.

and

> Unmanaged versions
> Your point is covered above. We do not accept hard coded urls. 

This is incorrect. See [`sublimetext2.plugin/install.sh`](https://github.com/folkswithhats/fedy/blob/master/plugins/sublimetext2.plugin/install.sh) as an example (I haven't gone through each and every plugin, there may be more).

Additionally, versions aren't the only thing that can change. URLs, file names, package structure can all change too - any one of these would immediately break a plugin for all users, whereas bundling a specific version of the `target` (Sublime Text v2.0.2 in this case) within a `fedy-plugin-*` system package would ensure that all users are always going to be able to successfully install the software when they run `dnf install fedy-plugin-*` (or trigger it through the GUI application). Another thing to consider is the reliability of the server that the `target` is hosted on - what if it is taken down, has issues, or is otherwise made unavailable? Sure, these issues can arise from a repository-based `target`, but that's the trade off for the convenience of using a repository.

> If the website has changed the download URL, then surely we do need to update the plugin but that is a small update. If they do not provide dnf streams or internal software updates then there is very little we can do.

If a repo isn't available, why _not_ make it a saner, more stable experience, instead of requiring us to refactor our project based on external changes like a URL change?

---

> We can't do anything when there is no dnf means legally like in case of sublime text which does not provide repo stream nor allows you to create copr yourself.

I'm not aware of any licensing restriction which would prevent bundling the freely available archive within a system package. In fact, asking in IRC and emailing their sales team confirms that redistributing any unactivated version of Sublime Text is fine; meaning we could bundle it without worry.

---

> Earlier version of fedy used to run system update with every start, we can have an tweak plugin to add cornjob to update system every day or alternate day in the background.

This is a horrible idea. Do not update user's systems.

**Comment by satya164** [comment]
I don't understand your idea of a plugin. The most important part of a Fedy plugin is that it contains metadata for Fedy. A normal RPM package includes instructions for the system to install and remove items, but not any metadata Fedy can use. It cannot be a Fedy plugin, unless you actually include the metadata or make Fedy understand RPM packages (and figure out a way to actually include an icon inside RPM package which fedy can use without having extract the RPM), in which case, every RPM package becomes a plugin.
 
> Fedy actually reinvents the wheel quite a bit, and has several problems:

It definitely does, but Fedy was never meant to be an app store or a package manager. It is just a tool which makes it easy to run some scripts on your system.

> We had to implement our own caching system

There's no caching system. It's a extremely simple convention to download files via wget to fedy's cache folder. Unlike any caching system.

> We are currently discussing methods of dependency management

Where? The only post I saw is when multiple packages are using the same repo. That's not dependency management. RPM packages don't solve this problem.

> We have several users who run into issues using workarounds Fedy has implemented, like rm inconsistencies, additional prompts for installation

The prompt is something Fedy has implemented. It's not a result of re-inventing the wheel.

> Plugins rely on external sources that can easily be changed, go down, or otherwise break

Yes, as I said previously, it's a tradeoff for maintainability. It's impossible for such a small team to keep track of all the packages and update every package version every time they change.

And this problem is not the result of re-inventing the wheel. Plugins can use hardcoded versions, but automatic version resolution is something I came up with. You don't need RPM packages to solve this problem.

> We have our own, custom solution for plugin structure, with no formal documentation or guidelines

I think the plugin system is extremely simple and straightforward to understand.

We had [docs](https://github.com/folkswithhats/fedy/wiki/Plugins-support) for previous version of Fedy which was arguably unnecessarily complex than it needed to be. We don't have docs now because nobody worked on it.

Using RPM packages doesn't solve this problem. You still need to write documentation and guidelines.

> Providing packages gives more control to the user

Has this ever been an issue? This is not Fedy's goal.

> Providing packages makes it easier for both users (easy to understand installation process) and developers (easy to debug).

I don''t think users care. And for developers, it's easier to debug shell scripts rather than packages.

> This is incorrect. See sublimetext2.plugin/install.sh as an example (I haven't gone through each and every plugin, there may be more).

Yes, there are few plugins where I was not able to figure out a way to resolve latest version, so they use hardcoded URL. But that doesn't make them cannon.

> If a repo isn't available, why not make it a saner, more stable experience, instead of requiring us to refactor our project based on external changes like a URL change?

Refactor and bugfix have very different meanings. But again, this is a tradeoff. It is like this because it's easier to maintain for us.

> I'm not aware of any licensing restriction which would prevent bundling the freely available archive within a system package.

I'm not sure about Sublime, but licenses can prohibit re-distribution of a software. And packaging means re-distributing. Anyway, this has workarounds.

> Earlier version of fedy used to run system update with every start, we can have an tweak plugin to add cornjob to update system every day or alternate day in the background.

@Abhinav1217 Hmm, not sure where you got the idea, but a core principle of Fedy has been to never change anything in the system unless the user specifically asks to do so. It never ran system update on every start. And it should never do this.

> What does gnome-software has to do with what this proposal is suggesting? Fedy plugins, currently a bundle of shell scripts, do not gain any of the benefits that system packages do from gnome-software .

Gnome Software does exactly what you are suggesting to do.

> .spec files are really not that difficult to understand or write -- speaking objectively.

Python is not difficult to understand or write. But I cannot write python beyond some super basic programs. Easy or hard depends on the knowledge level.

> Yes, the structure of the document can take a few minutes to get the basics of, but the commands you're writing are basically bash.

Some parts of spec files resemble and even use bash commands. But there are several places where there are different way to do the same thing. I can't get into details, but it's important to know the differences. A familiar syntax with differences is worse than a non-familiar syntax where it's obvious that you can't use the current way.

> Good contributing docs (which we are currently lacking anyways) will make the structured format of .spec files a non-issue.

Do you think we can/should document the format of spec files which is insanely huge and already has lots of documentation?

>  If we provide system packages, we know that they will always work the same for everyone - there will be no inconsistencies between the development testing experience and between users.

That's simply not true. It highly depends on environment configuration. I've never seen an issue arising due to different package versions. Not saying packages aren't better, but we have never faced this issue.

---

Anyway, enough bike-shedding, we had discussed it earlier, I am not saying that packages are not superior, they definitely are, but they are complex at the same time, understandable because they solve a lot more problems than what Fedy plugins need to.

You need to ask these questions to yourself,

Does the hugely increased maintenance cost really justifies the move? You can automate, but how much? You still need to keep track of every software, bump the version numbers when they change. There is a reason there are dedicated package maintainers who maintain a limited number of software.

You have time to work on it right now. But do you guarantee that it'll be true in future? Will the current contributors, who are actively contributing to Fedy, still be able to contribute without having to spend significant time learning about spec files and packaging, or they will get discouraged never/rarely contribute back? Does the new system make it more difficult for new contributors?

**Comment by Abhinav1217** [comment]
The increased complexity in maintenance and increased learning curve for new contributors are enough reason for me to  reject this proposal. Bash scripts works. A lot of problems will be solved if we get drpms. having drpms will reduce the size of updates to very small size. We can increase the frequency of updates to keep plugins updated.

**Comment by angystardust** [comment]
IMHO a more flexible and reliable alternative to bash script would be ansible playbooks (maybe runned by a shell wrapper). An example for installing the "Archive formats":
```
    - name: Install various archive formats (Fedy)
      package:
        name: '{{ item }}'
        state: latest
      with_items:
        - cabextract
        - lzip
        - p7zip
        - p7zip-plugins
        - unrar
```
P.S.: using Ansible playbooks to configure the system is the way redhat/fedora guys are pointing to.
Take a look at this very interesting project:
https://github.com/cockpit-project/system-api-roles

**Comment by stephenfin** [comment]
I wholeheartedly agree with this idea. I would argue that, while a focus on contributor experience *is* important, the focus on the end user experience is *more* important. In addition, I think idea that having to learn how to write spec files is a barrier to entry is a bit of straw man argument: not only is it a very useful skill but since when is (safe) shell scripting a piece of cake (the amount of `rm -rf /$undefined_var` statements I see in code is crazy)? Happy to help where I can.

**Comment by bddenhartog** [comment]
@stephenfin @angrystardust I've actually made a separate github organization to expand upon the ideas purported here in a cross-distro fashion. Feel free to contact me on telegram or twitter: @bddenhartog.

**Comment by sudoforge** [comment]
I'll be working on this shortly after #556 is closed.

**Comment by sudoforge** [comment]
Closing in favor of #607.
